src/main/scala/mesosphere/marathon/state/AppDefinition.scala

```diff
@@ -365,6 +366,38 @@ case class AppDefinition(
     val baseId = id.canonicalPath(base)
     copy(id = baseId, dependencies = dependencies.map(_.canonicalPath(baseId)))
   }
+
+  def portAssignments(task: Task): Option[Seq[PortAssignment]] = {
+    def fromIpAddress: Option[Seq[PortAssignment]] = ipAddress.flatMap {
+      case IpAddress(_, _, DiscoveryInfo(appPorts)) =>
+        for {
+          launched <- task.launched
+          effectiveIpAddress <- task.effectiveIpAddress(this)
+        } yield appPorts.zip(launched.hostPorts).zipWithIndex.map {
+          case ((appPort, hostPort), portIndex) =>
+            PortAssignment(Some(appPort.name), portIndex, effectiveIpAddress, hostPort)
+        }.toList
+    }
+
+    def fromPortMappings: Option[Seq[PortAssignment]] =
+      for {
+        pms <- portMappings
+        launched <- task.launched
+      } yield pms.zip(launched.hostPorts).zipWithIndex.map {
+        case ((portMapping, hostPort), portIndex) =>
+          PortAssignment(portMapping.name, portIndex, task.agentInfo.host, hostPort)
+      }.toList
+
+    def fromPortDefinitions: Option[Seq[PortAssignment]] = task.launched.map { launched =>
+      portDefinitions.zip(launched.hostPorts).zipWithIndex.map {
+        case ((portDefinition, hostPort), portIndex) =>
+          PortAssignment(portDefinition.name, portIndex, task.agentInfo.host, hostPort)
+      }
+    }
+
+    if (ipAddress.isDefined) fromIpAddress // If an IP-per-Task was requested, don't fall back
```

Comment:
but maybe something like `if (ipAddress.isDefined)` is good. Maybe it would even be nice if the top-level thing here was kind of self-explanatory, e.g. we use discoveryInfos for ip-per-container, we use portMappings for docker bridge mode and we use portDefinitions for docker HOST mode and non-docker apps.